src/vmm/src/devices/virtio/block/device.rs

```diff
@@ -1,7 +1,202 @@
 // Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
 
+use std::sync::atomic::AtomicU32;
+use std::sync::Arc;
+
+use event_manager::{EventOps, Events, MutEventSubscriber};
+use snapshot::Persist;
+use utils::eventfd::EventFd;
+
+use super::persist::{BlockConstructorArgs, BlockState};
 use super::virtio_block::device::{VirtioBlock, VirtioBlockConfig};
+use super::BlockError;
+use crate::devices::virtio::device::VirtioDevice;
+use crate::devices::virtio::queue::Queue;
+use crate::devices::virtio::{ActivateError, TYPE_BLOCK};
+use crate::rate_limiter::BucketUpdate;
+use crate::vstate::memory::GuestMemoryMmap;
 
-pub type Block = VirtioBlock;
 pub type BlockConfig = VirtioBlockConfig;
+
+#[derive(Debug)]
+pub enum Block {
+    Virtio(VirtioBlock),
+}
+
+impl Block {
+    pub fn new(config: BlockConfig) -> Result<Block, BlockError> {
+        Ok(Self::Virtio(VirtioBlock::new(config)?))
+    }
+
+    pub fn config(&self) -> BlockConfig {
+        match self {
+            Self::Virtio(b) => b.config(),
+        }
+    }
+
+    pub fn update_disk_image(&mut self, disk_image_path: String) -> Result<(), BlockError> {
+        match self {
+            Self::Virtio(b) => b.update_disk_image(disk_image_path),
+        }
+    }
+
+    pub fn update_rate_limiter(&mut self, bytes: BucketUpdate, ops: BucketUpdate) {
+        match self {
+            Self::Virtio(b) => b.update_rate_limiter(bytes, ops),
```

Comment:
Now `Block` is aware of differences between implementations (virtio vs vhost-user). Is that what we want or would we like to encapsulate everything inside its variants?

Reply:
How would 'encapsulate everything inside its variants' look like?

Reply:
I never considered that `vhost-user` will implement functions that do not make sense for it. I think we should keep each block type with only methods needed for it and the `Block` can handle all nuances. Also it is easier to return `BlockError` from `Block` rather than from `virtio` or `vhost-user` blocks.

Reply:
`update_rate_limiter` does return an error. It is added with vhost-user block addition in later commits.

Reply:
could you please bring it over to this commit?

Reply:
done